qiskit_ode/signals/signals.py

```diff
@@ -282,24 +359,27 @@ def start_time(self) -> float:
         """
         return self._start_time
 
-    def envelope_value(self, t: float = 0.0) -> complex:
-
-        idx = Array((t - self._start_time) // self._dt, dtype=int)
-
-        return Array(self._samples.data[idx.data])
+    def envelope(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Envelope. If ``t`` is before (resp. after) the start (resp. end) of the definition of
+        the ``DiscreteSignal```, this will return the start value (resp. end value).
+        """
+        idx = np.clip(
+            Array((t - self._start_time) // self._dt, dtype=int), 0, len(self._samples) - 1
+        )
+        return self._samples[idx.data]
 
-    def value(self, t: float = 0.0) -> Array:
+    def complex_value(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
         """Return the value of the signal at time t."""
-        arg = 1j * 2 * np.pi * self.carrier_freq * t + 1j * self.phase
-        return np.real(self.envelope_value(t) * np.exp(arg))
+        arg = self._carrier_arg * t + self._phase_arg
+        return self.envelope(t) * np.exp(arg)
 
     def conjugate(self):
-        return PiecewiseConstant(
+        return DiscreteSignal(
             dt=self._dt,
             samples=np.conjugate(self._samples),
             start_time=self._start_time,
             duration=self.duration,
-            carrier_freq=-self.carrier_freq,
+            carrier_freq=self.carrier_freq,
             phase=-self.phase,
         )
```

Comment:
How do we interpret `conjugate` for a real signal?

Reply:
This is a good point:
- In this PR, for a signal `Re[f(t)exp(i(wt + phi))]`, conjugation maps the signal to one implementing the function `Re[conj(f(t))exp(i(wt - phi))]`; basically taking the conjugate of the envelope, and conceptually including the phase as part of that. 
- I chose this as it is useful in implementing multiplication of signals, as we do really need to just take the conjugate of the envelope/phase. Also, since signals are explicitly real now, also taking `w` -> `-w` would result in the same real-valued function.
- However, this behaviour is definitely weird with the method `complex_value`, as with this definition `sig2 = sig1.conjugate()` won't satisfy `sig2.complex_value(t) == np.conj(sig1.complex_value(t))`.

Some options: 
- We could change `.conjugate()` to `.conjugate_envelope()` to be explicit about this. 
- Or `.conjugate()` could be expanded to have optional arguments about which parameters in the complex representation are being "conjugated". E.g. it could have a signature: 
```
def conjugate(self, envelope=True, carrier_freq=True, phase=True):
```

My preference would be the latter option.